Modules/TOF/src/TrendingCalibLHCphase.cxx

```diff
@@ -0,0 +1,206 @@
+#include "QualityControl/QcInfoLogger.h"
+#include "QualityControl/DatabaseInterface.h"
+#include "QualityControl/MonitorObject.h"
+#include "QualityControl/Reductor.h"
+#include "QualityControl/RootClassFactory.h"
+#include "QualityControl/RepoPathUtils.h"
+#include "QualityControl/ActivityHelpers.h"
+#include "CCDB/BasicCCDBManager.h"
+#include "QualityControl/UserCodeInterface.h"
+#include "DetectorsCalibration/Utils.h"
+#include "TOFBase/Utils.h"
+#include "TOFCalibration/LHCClockCalibrator.h"
+#include "TOFBase/Geo.h"
+#include "CCDB/CcdbApi.h"
+
+#include <TH1.h>
+#include <TMath.h>
+#include <TH2.h>
+#include <TCanvas.h>
+#include <TPaveText.h>
+#include <TDatime.h>
+#include <TGraphErrors.h>
+#include <TProfile.h>
+#include <TPoint.h>
+
+#include <boost/property_tree/ptree.hpp>
+
+using namespace std;
+using namespace o2::quality_control;
+using namespace o2::quality_control::core;
+using namespace o2::quality_control::postprocessing;
+using namespace o2::quality_control_modules::tof;
+
+using LHCphase = o2::dataformats::CalibLHCphaseTOF;
+
+void TrendingCalibLHCphase::configure(const boost::property_tree::ptree& config)
+{
+  mConfig = TrendingConfigTOF(getID(), config);
+  mHost = config.get<std::string>("qc.postprocessing." + getID() + ".dataSourceURL");
+}
+
+void TrendingCalibLHCphase::initialize(Trigger, framework::ServiceRegistryRef)
+{
+  mCdbApi.init(mHost);
+
+  // Preparing data structure of TTree
+  mTrend = std::make_unique<TTree>();
+  mTrend->SetName(PostProcessingInterface::getName().c_str());
+  mTrend->Branch("runNumber", &mMetaData.runNumber);
+  mTrend->Branch("time", &mTime);
+  mTrend->Branch("phase", &mPhase);
+
+  for (const auto& source : mConfig.dataSources) {
+    std::unique_ptr<Reductor> reductor(root_class_factory::create<Reductor>(source.moduleName, source.reductorName));
+    mTrend->Branch(source.name.c_str(), reductor->getBranchAddress(), reductor->getBranchLeafList());
+    mReductors[source.name] = std::move(reductor);
+  }
+  getObjectsManager()->startPublishing(mTrend.get());
+}
+
+void TrendingCalibLHCphase::update(Trigger t, framework::ServiceRegistryRef services)
+{
+  auto& ccdb = services.get<repository::DatabaseInterface>();
+
+  trendValues(t, ccdb);
+  generatePlots();
+}
+
+void TrendingCalibLHCphase::finalize(Trigger t, framework::ServiceRegistryRef)
+{
+  generatePlots();
+}
+
+void TrendingCalibLHCphase::trendValues(const Trigger& t, repository::DatabaseInterface& ccdb)
+{
+  mTime = activity_helpers::isLegacyValidity(t.activity.mValidity)
+            ? t.timestamp / 1000
+            : t.activity.mValidity.getMax() / 1000; // ROOT expects seconds since epoch.
+  mMetaData.runNumber = t.activity.mId;
+
+  mPhase = 0.;
+
+  map<string, string> metadata; // can be empty
+
+  for (auto& dataSource : mConfig.dataSources) {
+
+    if (dataSource.type == "ccdb") {
+
+      auto calib_object = mCdbApi.retrieveFromTFileAny<LHCphase>(dataSource.path, metadata, -1);
```

Comment:
Since very recently, you can just use `UserCodeInterface::retrieveConditionAny` (an inherited member of TrendingCalibLHCPhase) instead of using CcdbAPI directly.

Also, soon you should be able to write only a Reductor implementation to achieve this, as it is demonstrated in `testReductor.cxx` in https://github.com/AliceO2Group/QualityControl/pull/2110.

Reply:
Hi @knopers8,
thanks for pointing this out! 
I am using now directly the member ``UserCodeInterface::retrieveConditionAny``, also I will keep an eye on the new Reductor implementation